classes/local/store/s3/client.php

```diff
@@ -472,19 +462,30 @@ private function generate_presigned_url_cloudfront($contenthash, $headers = [],
         if ($nicefilename) {
             $key .= $this->get_nice_filename($headers);
         }
+        $resource = $this->config->cloudfrontresourcedomain . '/' . $key;
+        // This is the id of the Cloudfront key pair you generated.
+        $keypairid = $this->config->cloudfrontkeypairid;
 
-        $resourcekey = $this->config->cloudfrontresourcedomain . '/' . $key;
-        $signingparameters = [
-            'url' => $resourcekey,
-            'expires' => $expires,
-            'key_pair_id' => $this->config->cloudfrontkeypairid,
-            'private_key' => realpath($this->config->cloudfrontprivatekeypemfilepathname),
-        ];
-        $signedurlcannedpolicy = $client->getSignedUrl($signingparameters);
-        $signedurl = (string)$signedurlcannedpolicy;
-
-        /* $headers[] = 'Location:"' . $signedurl . '"'; // This may cause loss of headers (etag for example). */
-        return $signedurl;
+        $json = '{"Statement":[{"Resource":"' . $resource . '","Condition":{"DateLessThan":{"AWS:EpochTime":' . $expires . '}}}]}';
+
+        // Create the private key.
+        $key = openssl_get_privatekey($this->config->cloudfrontprivatekey);
+        if (!$key) {
+            throw new \Exception('Loading private key failed');
```

Comment:
would this be better as a moodle_exception? and below

Reply:
also 'Loading private key failed' doesn't mention what plugin or config we are dealing with, maybe something like 'tool_object could not load cloudfront signing key'

Reply:
updated